File: WordPress/Classes/Models/Post.m
#import "Post.h"
#import "Media.h"
#import "PostCategory.h"
#import "Coordinate.h"
#import "NSDate+StringFormatting.h"
#import "NSMutableDictionary+Helpers.h"
#import "NSString+Helpers.h"
#import "ContextManager.h"
#import <WordPressShared/NSString+XMLExtensions.h>

NSString * const PostTypeDefaultIdentifier = @"post";

@interface Post()
@property (nonatomic, strong) NSString *storedContentPreviewForDisplay;
@end

@implementation Post

@dynamic commentCount;
@dynamic likeCount;
@dynamic geolocation;
@dynamic tags;
@dynamic postType;
@dynamic postFormat;
@dynamic latitudeID;
@dynamic longitudeID;
@dynamic publicID;
@dynamic categories;
@synthesize specialType;
@synthesize storedContentPreviewForDisplay;

#pragma mark - NSManagedObject subclass methods

- (void)awakeFromFetch
{
    [super awakeFromFetch];

    [self buildContentPreview];
}

- (void)didTurnIntoFault
{
    [super didTurnIntoFault];

    self.specialType = nil;
}

- (void)willSave
{
    [super willSave];
    if ([self isDeleted]) {
        return;
    }
    [self buildContentPreview];
}


#pragma mark -

- (void)buildContentPreview
{
    NSString *str = self.mt_excerpt;
    if ([str length]) {
        str = [NSString makePlainText:str];
    } else {
        str = [BasePost summaryFromContent:self.content];
    }
    self.storedContentPreviewForDisplay = str ? str : @"";
}

- (NSString *)categoriesText
{
    return [[[self.categories valueForKey:@"categoryName"] allObjects] componentsJoinedByString:@", "];
}

- (NSString *)postFormatText
{
    return [self.blog postFormatTextFromSlug:self.postFormat];
}

- (void)setPostFormatText:(NSString *)postFormatText
{
    __block NSString *format = nil;
    [self.blog.postFormats enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        if ([obj isEqual:postFormatText]) {
            format = (NSString *)key;
            *stop = YES;
        }
    }];
    self.postFormat = format;
}

- (void)setCategoriesFromNames:(NSArray *)categoryNames
{
    [self removeCategories:self.categories];
    NSMutableSet *categories = nil;

    for (NSString *categoryName in categoryNames) {
        NSSet *results = [self.blog.categories filteredSetUsingPredicate:[NSPredicate predicateWithFormat:@"categoryName = %@", categoryName]];
        if (results && (results.count > 0)) {
            if (categories == nil) {
                categories = [NSMutableSet setWithSet:results];
            } else {
                [categories unionSet:results];
            }
        }
    }

    if (categories && (categories.count > 0)) {
        [self addCategories:categories];
    }
}

- (BOOL)hasSiteSpecificChanges
{
    if ([super hasSiteSpecificChanges]) {
        return YES;
    }

    Post *original = (Post *)self.original;

    if ((self.postFormat != original.postFormat) && (![self.postFormat isEqual:original.postFormat])) {
        return YES;
    }

    if (![self.categories isEqual:original.categories]) {
        return YES;
    }

    return NO;
}

- (BOOL)hasCategories
{
    if ([self.categories count] > 0) {
        return true;
    }

    return false;
}

- (BOOL)hasTags
{
    if ([[self.tags trim] length] > 0) {
        return true;
    }

    return false;
}

#pragma mark - Unsaved changes

- (BOOL)hasLocalChanges
{
    if ([super hasLocalChanges]) {
        return YES;
    }
    
    Post *original = (Post *)self.original;
    if (!original) {
        return NO;
    }
    
    if (([self.tags length] != [original.tags length]) && (![self.tags isEqual:original.tags])) {
        return YES;
    }
    
    CLLocationCoordinate2D coord1 = self.geolocation.coordinate;
    CLLocationCoordinate2D coord2 = original.geolocation.coordinate;
    if ((coord1.latitude != coord2.latitude) || (coord1.longitude != coord2.longitude)) {
        return YES;
    }
    
    return NO;
}

#pragma mark - WPPostContentViewProvider Methods

- (NSString *)titleForDisplay
{
    NSString *title = [self.postTitle stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] ?: @"";
    title = [title stringByDecodingXMLCharacters];
    if ([title length] == 0 && [[self contentPreviewForDisplay] length] == 0 && ![self hasRemote]) {
        title = NSLocalizedString(@"(no title)", @"Let's a user know that a local draft does not have a title.");
    }
    return title;
}

- (NSString *)contentPreviewForDisplay
{
    if (self.storedContentPreviewForDisplay == nil) {
        [self buildContentPreview];
    }
    return self.storedContentPreviewForDisplay;
}

- (NSString *)statusForDisplay
{
    NSString *statusString = [self statusTitle];
    if ([self.status isEqualToString:PostStatusPublish] || [self.status isEqualToString:PostStatusDraft]) {
        statusString = nil;
    }
    if (![self hasRemote]) {
        NSString *localOnly = NSLocalizedString(@"Local", @"A status label for a post that only exists on the user's iOS device, and has not yet been published to their blog.");
        if (statusString) {
            statusString = [NSString stringWithFormat:@"%@, %@", statusString, localOnly];
        } else {
            statusString = localOnly;
        }
    }
    return statusString;
}

- (NSURL *)featuredImageURLForDisplay
{
    NSURL *url = [NSURL URLWithString:self.pathForDisplayImage];
    return url;
}

- (NSInteger)numberOfComments
{
    if (self.commentCount) {
        return [self.commentCount integerValue];
    }
    return 0;
}

- (NSInteger)numberOfLikes
{
    if (self.likeCount) {
        return [self.likeCount integerValue];
    }
    return 0;
}


@end

File: WordPress/Classes/Models/Post.swift
import Foundation
import CoreData

@objc(Post)
class Post: AbstractPost {
    @objc static let typeDefaultIdentifier = "post"

    struct Constants {
        static let publicizeIdKey = "id"
        static let publicizeValueKey = "value"
        static let publicizeKeyKey = "key"
        static let publicizeDisabledValue = "1"
        static let publicizeEnabledValue = "0"
    }

    enum PublicizeMetadataSkipPrefix: String {
        case keyring = "_wpas_skip_"
        case connection = "_wpas_skip_publicize_"

        /// Determines the prefix type from the given key.
        ///
        /// - Parameter key: String.
        /// - Returns: A `PublicizeMetadataSkipPrefix` value, or nil if nothing matched.
        static func prefix(of key: String) -> PublicizeMetadataSkipPrefix? {
            // try to match the `keyring` format first, since it's a substring of the `connection` format.
            guard key.hasPrefix(Self.keyring.rawValue) else {
                return nil
            }
            return key.hasPrefix(Self.connection.rawValue) ? .connection : .keyring
        }
    }

    // MARK: - NSManagedObject

    override class func entityName() -> String {
        return "Post"
    }

    // MARK: - Format

    @objc func postFormatText() -> String? {
        return blog.postFormatText(fromSlug: postFormat)
    }

    @objc func setPostFormatText(_ postFormatText: String) {

        assert(blog.postFormats is [String: String])
        guard let postFormats = blog.postFormats as? [String: String] else {
            DDLogError("Expected blog.postFormats to be \(String(describing: [String: String].self)).")
            return
        }

        var formatKey: String?

        for (key, value) in postFormats {
            if value == postFormatText {
                formatKey = key
                break
            }
        }

        postFormat = formatKey
    }

    // MARK: - Categories

    /// Returns categories as a comma-separated list
    ///
    @objc func categoriesText() -> String {

        guard let allStrings = categories?.map({ return $0.categoryName as String }) else {
            return ""
        }

        let orderedStrings = allStrings.sorted { (categoryName1, categoryName2) -> Bool in
            return categoryName1.localizedCaseInsensitiveCompare(categoryName2) == .orderedAscending
        }

        return orderedStrings.joined(separator: ", ")
    }

    /// Set the categories for a post
    ///
    /// - Parameter categoryNames: a `NSArray` with the names of the categories for this post. If
    ///                     a given category name doesn't exist it's ignored.
    ///
    @objc func setCategoriesFromNames(_ categoryNames: [String]) {

        var newCategories = Set<PostCategory>()

        for categoryName in categoryNames {

            assert(blog.categories is Set<PostCategory>)
            guard let blogCategories = blog.categories as? Set<PostCategory> else {
                DDLogError("Expected blog.categories to be \(String(describing: Set<PostCategory>.self)).")
                return
            }

            let matchingCategories = blogCategories.filter({ return $0.categoryName == categoryName })

            if matchingCategories.count > 0 {
                newCategories = newCategories.union(matchingCategories)
            }
        }

        categories = newCategories
    }

    // MARK: - Sharing

    @objc func canEditPublicizeSettings() -> Bool {
        return !self.hasRemote() || self.status != .publish
    }

    // MARK: - PublicizeConnections

    @objc func publicizeConnectionDisabledForKeyringID(_ keyringID: NSNumber) -> Bool {
        let isKeyringEntryDisabled = disabledPublicizeConnections?[keyringID]?[Constants.publicizeValueKey] == Constants.publicizeDisabledValue

        // try to check in case there's an entry for the PublicizeConnection that's keyed by the connectionID.
        guard let connections = blog.connections as? Set<PublicizeConnection>,
              let connection = connections.first(where: { $0.keyringConnectionID == keyringID }),
              let existingValue = disabledPublicizeConnections?[connection.connectionID]?[Constants.publicizeValueKey] else {
            // fall back to keyringID if there is no such entry with the connectionID.
            return isKeyringEntryDisabled
        }

        let isConnectionEntryDisabled = existingValue == Constants.publicizeDisabledValue
        return isConnectionEntryDisabled || isKeyringEntryDisabled
    }

    @objc func enablePublicizeConnectionWithKeyringID(_ keyringID: NSNumber) {
        // if there's another entry keyed by connectionID references to the same connection,
        // we need to make sure that the values are kept in sync.
        if let connections = blog.connections as? Set<PublicizeConnection>,
           let connection = connections.first(where: { $0.keyringConnectionID == keyringID }),
           let _ = disabledPublicizeConnections?[connection.connectionID] {
            enablePublicizeConnection(keyedBy: connection.connectionID)
        }

        enablePublicizeConnection(keyedBy: keyringID)
    }

    @objc func disablePublicizeConnectionWithKeyringID(_ keyringID: NSNumber) {
        // if there's another entry keyed by connectionID references to the same connection,
        // we need to make sure that the values are kept in sync.
        if let connections = blog.connections as? Set<PublicizeConnection>,
           let connectionID = connections.first(where: { $0.keyringConnectionID == keyringID })?.connectionID,
           let _ = disabledPublicizeConnections?[connectionID] {
            disablePublicizeConnection(keyedBy: connectionID)

            // additionally, if the keyring entry doesn't exist, there's no need create both formats.
            // we can just update the dictionary's key from connectionID to keyringID instead.
            if disabledPublicizeConnections?[keyringID] == nil,
               let updatedEntry = disabledPublicizeConnections?[connectionID] {
                disabledPublicizeConnections?.removeValue(forKey: connectionID)
                disabledPublicizeConnections?[keyringID] = updatedEntry
                return
            }
        }

        disablePublicizeConnection(keyedBy: keyringID)
    }

    /// Marks the Publicize connection with the given id as enabled.
    ///
    /// - Parameter id: The dictionary key for `disabledPublicizeConnections`.
    private func enablePublicizeConnection(keyedBy id: NSNumber) {
        guard var connection = disabledPublicizeConnections?[id] else {
            return
        }

        // if the auto-sharing settings is not yet synced to remote,
        // we can just remove the entry since all connections are enabled by default.
        guard let _ = connection[Constants.publicizeIdKey] else {
            _ = disabledPublicizeConnections?.removeValue(forKey: id)
            return
        }

        connection[Constants.publicizeValueKey] = Constants.publicizeEnabledValue
        disabledPublicizeConnections?[id] = connection
    }

    /// Marks the Publicize connection with the given id as disabled.
    ///
    /// - Parameter id: The dictionary key for `disabledPublicizeConnections`.
    private func disablePublicizeConnection(keyedBy id: NSNumber) {
        if let _ = disabledPublicizeConnections?[id] {
            disabledPublicizeConnections?[id]?[Constants.publicizeValueKey] = Constants.publicizeDisabledValue
            return
        }

        if disabledPublicizeConnections == nil {
            disabledPublicizeConnections = [NSNumber: [String: String]]()
        }

        disabledPublicizeConnections?[id] = [Constants.publicizeValueKey: Constants.publicizeDisabledValue]
    }

    // MARK: - Comments

    @objc func numberOfComments() -> Int {
        return commentCount?.intValue ?? 0
    }

    // MARK: - Likes

    @objc func numberOfLikes() -> Int {
        return likeCount?.intValue ?? 0
    }

    // MARK: - AbstractPost

    override func hasCategories() -> Bool {
        categories?.isEmpty == false
    }

    override func hasTags() -> Bool {
        tags?.trim().isEmpty == false
    }

    override func authorForDisplay() -> String? {
        author ?? blog.account?.displayName
    }

    // MARK: - BasePost

    override func contentPreviewForDisplay() -> String {
        if let excerpt = mt_excerpt, excerpt.count > 0 {
            if let preview = PostPreviewCache.shared.excerpt[excerpt] {
                return preview
            }
            let preview = excerpt.makePlainText()
            PostPreviewCache.shared.excerpt[excerpt] = preview
            return preview
        } else if let content {
            if let preview = PostPreviewCache.shared.content[content] {
                return preview
            }
            let preview = content.summarized()
            PostPreviewCache.shared.content[content] = preview
            return preview
        } else {
            return ""
        }
    }

    override func titleForDisplay() -> String {
        var title = postTitle?.trimmingCharacters(in: CharacterSet.whitespaces) ?? ""
        title = title
            .stringByDecodingXMLCharacters()
            .strippingHTML()

        if title.count == 0 && contentPreviewForDisplay().count == 0 && !hasRemote() {
            title = NSLocalizedString("(no title)", comment: "Lets a user know that a local draft does not have a title.")
        }

        return title
    }
}

private final class PostPreviewCache {
    static let shared = PostPreviewCache()

    let excerpt = Cache<String, String>()
    let content = Cache<String, String>()
}

private final class Cache<Key: Hashable, Value> {
    private let lock = NSLock()
    private var dictionary: [Key: Value] = [:]

    subscript(key: Key) -> Value? {
        get { lock.withLock { dictionary[key] } }
        set { lock.withLock { dictionary[key] = newValue } }
    }
}
